pkg/detector/detector.go

```diff
@@ -614,51 +641,95 @@ func (d *ResourceDetector) GetUnstructuredObject(objectKey keys.ClusterWideKey)
+		if err := d.Client.Update(context.TODO(), policy); err != nil {
+			return id, err
+		}
+	}
+
+	return id, nil
+}
+
 // ClaimPolicyForObject set policy identifier which the object associated with.
-func (d *ResourceDetector) ClaimPolicyForObject(object *unstructured.Unstructured, policyNamespace, policyName, policyUID string) error {
+func (d *ResourceDetector) ClaimPolicyForObject(object *unstructured.Unstructured, policy *policyv1alpha1.PropagationPolicy) (string, error) {
+	policyID, err := d.getPropagationPolicyID(policy)
+	if err != nil {
+		klog.Errorf("Get PropagationPolicy(%s/%s) ID error:%v", policy.Namespace, policy.Name, err)
+		return "", err
+	}
+
 	objLabels := object.GetLabels()
 	if objLabels == nil {
 		objLabels = make(map[string]string)
 	} else if len(objLabels) > 0 {
 		// object has been claimed, don't need to claim again
 		if !excludeClusterPolicy(objLabels) &&
-			objLabels[policyv1alpha1.PropagationPolicyNamespaceLabel] == policyNamespace &&
-			objLabels[policyv1alpha1.PropagationPolicyNameLabel] == policyName {
-			return nil
+			objLabels[policyv1alpha1.PropagationPolicyIDLabel] == policyID {
+			return policyID, nil
 		}
 	}
 
-	objLabels[policyv1alpha1.PropagationPolicyNamespaceLabel] = policyNamespace
-	objLabels[policyv1alpha1.PropagationPolicyNameLabel] = policyName
-	objLabels[policyv1alpha1.PropagationPolicyUIDLabel] = policyUID
+	// Delete following line when release-1.9
+	delete(objLabels, policyv1alpha1.PropagationPolicyUIDLabel)
+	objLabels[policyv1alpha1.PropagationPolicyNamespaceLabel] = policy.Namespace
+	objLabels[policyv1alpha1.PropagationPolicyNameLabel] = policy.Name
+	objLabels[policyv1alpha1.PropagationPolicyIDLabel] = policyID
 
 	objectAnnotations := object.GetAnnotations()
 	if objectAnnotations == nil {
 		objectAnnotations = make(map[string]string)
 	}
-	objectAnnotations[policyv1alpha1.PropagationPolicyNamespaceAnnotation] = policyNamespace
-	objectAnnotations[policyv1alpha1.PropagationPolicyNameAnnotation] = policyName
+	objectAnnotations[policyv1alpha1.PropagationPolicyNamespaceAnnotation] = policy.Namespace
+	objectAnnotations[policyv1alpha1.PropagationPolicyNameAnnotation] = policy.Name
 
 	objectCopy := object.DeepCopy()
 	objectCopy.SetLabels(objLabels)
 	objectCopy.SetAnnotations(objectAnnotations)
-	return d.Client.Update(context.TODO(), objectCopy)
+	return policyID, d.Client.Update(context.TODO(), objectCopy)
 }
 
-// ClaimClusterPolicyForObject set cluster identifier which the object associated with.
-func (d *ResourceDetector) ClaimClusterPolicyForObject(object *unstructured.Unstructured, policyName, policyUID string) error {
-	claimedName := util.GetLabelValue(object.GetLabels(), policyv1alpha1.ClusterPropagationPolicyLabel)
+func (d *ResourceDetector) getClusterPropagationPolicyID(policy *policyv1alpha1.ClusterPropagationPolicy) (string, error) {
+	id := util.GetLabelValue(policy.GetLabels(), policyv1alpha1.ClusterPropagationPolicyIDLabel)
+	if id == "" {
```

Comment:
I believe we can also add this logic in the `webhook` to reduce unnecessary modifications to PP and decrease the number of times `Reconcile`.

Reply:
In the upgrading scenario, if the resources already exist in etcd, it will not trigger a webhook.

Reply:
I understand what you mean, but in my opinion, this logic should exist in two places. One is here and the other is in the webhook. They separately handle PPs that already exist and the creation of new PPs.

Reply:
I got it

Reply:
Yes, this way can reduce unnecessary `Reconcile`.

Reply:
Yes, agree, this logic is mainly used for migrating legacy policies from the previous version. 
We can safely remove it at the next release if we can do it in webhook.

Reply:
Let's do this in another PR.

Reply:
done, https://github.com/karmada-io/karmada/issues/4000#issuecomment-1776831548